src/main/io/serial_1wire.c

```diff
@@ -0,0 +1,235 @@
+ *  by Nathan Tsoi <nathan@vertile.com>
+ */
+
+#include <stdbool.h>
+
+#include "platform.h"
+
+#ifdef USE_SERIAL_1WIRE
+
+#include "drivers/gpio.h"
+#include "drivers/inverter.h"
+#include "drivers/light_led.h"
+#include "drivers/system.h"
+#include "io/serial_1wire.h"
+
+// Figure out esc clocks and pins, extrapolated from timer.c
+// Periphs could be pulled progmatically... but I'll leave that for another exercise
+#if defined(STM32F3DISCOVERY) && !(defined(CHEBUZZF3))
+const escHardware_t escHardware[ESC_COUNT] = {
+  { GPIOD, 12 },
+  { GPIOD, 13 },
+  { GPIOD, 14 },
+  { GPIOD, 15 },
+  { GPIOA, 1 },
+  { GPIOA, 2 }
+};
+#elif defined(CJMCU) || defined(EUSTM32F103RC) || defined(NAZE) || defined(OLIMEXINO) || defined(PORT103R)
+const escHardware_t escHardware[ESC_COUNT] = {
+  { GPIOA, 8 },
+  { GPIOA, 11 },
+  { GPIOB, 6 },
+  { GPIOB, 7 },
+  { GPIOB, 8 },
+  { GPIOB, 9 }
+};
+#elif CC3D
+const escHardware_t escHardware[ESC_COUNT] = {
+  { GPIOB, 9 },
+  { GPIOB, 8 },
+  { GPIOB, 7 },
+  { GPIOA, 8 },
+  { GPIOB, 4 },
+  { GPIOA, 2 }
+};
+#endif
+
+static void gpio_set_mode(GPIO_TypeDef* gpio, uint16_t pin, GPIO_Mode mode) {
+  gpio_config_t cfg;
+  cfg.pin = pin;
+  cfg.mode = mode;
+  cfg.speed = Speed_10MHz;
+  gpioInit(gpio, &cfg);
+}
+
+#ifdef STM32F10X
+static volatile uint32_t original_cr_mask, in_cr_mask, out_cr_mask;
+static __IO uint32_t *cr;
+static void gpio_prep_vars(uint16_t escIndex)
+{
+  GPIO_TypeDef *gpio = escHardware[escIndex].gpio;
+  uint32_t pinpos = escHardware[escIndex].pinpos;
+  // mask out extra bits from pinmode, leaving just CNF+MODE
+  uint32_t inmode = Mode_IPU & 0x0F;
+  uint32_t outmode = (Mode_Out_PP & 0x0F) | Speed_10MHz;
+  // reference CRL or CRH, depending whether pin number is 0..7 or 8..15
+  cr = &gpio->CRL + (pinpos / 8);
+  // offset to CNF and MODE portions of CRx register
+  uint32_t shift = (pinpos % 8) * 4;
+  // Read out current CRx value
+  original_cr_mask = in_cr_mask = out_cr_mask = *cr;
+  // Mask out 4 bits
+  in_cr_mask &= ~(0xF << shift);
+  out_cr_mask &= ~(0xF << shift);
+  // save current pinmode
+  in_cr_mask |= inmode << shift;
+  out_cr_mask |= outmode << shift;
+}
+
+static void gpioSetOne(uint16_t escIndex, GPIO_Mode mode) {
+  // reference CRL or CRH, depending whether pin number is 0..7 or 8..15
+  if (mode == Mode_IPU) {
+    *cr = in_cr_mask;
+    escHardware[escIndex].gpio->ODR |= (1U << escHardware[escIndex].pinpos);
+  }
+  else {
+    *cr = out_cr_mask;
+  }
+}
+#endif
+
+#define disable_hardware_uart  __disable_irq()
+#define enable_hardware_uart   __enable_irq()
+#define ESC_HI(escIndex)       ((escHardware[escIndex].gpio->IDR & (1U << escHardware[escIndex].pinpos)) != (uint32_t)Bit_RESET)
+#define RX_HI                  ((S1W_RX_GPIO->IDR & S1W_RX_PIN) != (uint32_t)Bit_RESET)
+#define ESC_SET_HI(escIndex)   escHardware[escIndex].gpio->BSRR = (1U << escHardware[escIndex].pinpos)
```

Comment:
Thanks for the code review @samguns! I did try that originally (e.g. https://github.com/nathantsoi/stm32-serial-1wire-passthrough/blob/master/src/serial_1wire.c#L128), but I couldn't get fast enough toggle speeds when switching modes on the 1wire line. Not sure why, I would think it would be plenty fast... but removing the loops that handle configuration of multiple pins in those libraries seems to have resolved the issue. Unfortunately, it did make this code a bit more messy: (ie. gpio_prep_vars)